fungible_conversion_proxy/src/lib.rs

```diff
@@ -0,0 +1,658 @@
+    pub fn new(oracle_account_id: AccountId, provider_account_id: AccountId) -> Self {
+        let owner_id = env::signer_account_id();
+        Self {
+            oracle_account_id,
+            provider_account_id,
+            owner_id,
+        }
+    }
+
+    pub fn set_oracle_account(&mut self, oracle: ValidAccountId) {
+        let signer_id = env::predecessor_account_id();
+        if self.owner_id == signer_id {
+            self.oracle_account_id = oracle.to_string();
+        } else {
+            panic!("ERR_PERMISSION");
+        }
+    }
+
+    pub fn get_oracle_account(&self) -> AccountId {
+        self.oracle_account_id.to_string()
+    }
+
+    pub fn set_provider_account(&mut self, oracle: ValidAccountId) {
+        let signer_id = env::predecessor_account_id();
+        if self.owner_id == signer_id {
+            self.provider_account_id = oracle.to_string();
+        } else {
+            panic!("ERR_PERMISSION");
+        }
+    }
+
+    pub fn get_provider_account(&self) -> AccountId {
+        self.provider_account_id.to_string()
+    }
+
+    pub fn set_owner(&mut self, owner: ValidAccountId) {
+        let signer_id = env::predecessor_account_id();
+        if self.owner_id == signer_id {
+            self.owner_id = owner.to_string();
+        } else {
+            panic!("ERR_PERMISSION");
+        }
+    }
+
+    #[private]
+    pub fn on_transfer_with_reference(
+        &self,
+        args: PayerSuppliedArgs,
+        payer: AccountId,
+        deposit: U128,
+        crypto_amount: U128,
+        crypto_fee_amount: U128,
+        change: U128,
+    ) -> String {
+        near_sdk::assert_self();
+
+        if near_sdk::is_promise_success() {
+            // Log success for indexing and payment detection
+            env::log(
+                &json!({
+                    "amount": args.amount,
+                    "currency": args.currency,
+                    "token_address": args.token_address,
+                    "fee_address": args.fee_address,
+                    "fee_amount": args.fee_amount,
+                    "max_rate_timespan": args.max_rate_timespan,
+                    "payment_reference": args.payment_reference,
+                    "to": args.to,
+                    "crypto_amount": crypto_amount,
+                    "crypto_fee_amount": crypto_fee_amount
+                })
+                .to_string()
+                .into_bytes(),
+            );
+            change.0.to_string() // return change for `ft_resolve_transfer` on the token contract
+        } else {
+            env::log(
+                format!(
+                    "Failed to transfer to account {}. Returning attached deposit of {} of token {} to {}",
+                    args.to, deposit.0, args.token_address, payer)
+                .as_bytes(),
+            );
+            deposit.0.to_string() // return full amount for `ft_resolve_transfer` on the token contract
+        }
+    }
+
+    #[private]
+    #[payable]
```

Comment:
```suggestion
```
Same here and for `rate_callback`, shouldn't we skip the payable attribute for ft payments? Maybe I miss something about Near, and then we can maybe skip `env::attached_deposit()`?